util/include/kvstream.h

```diff
@@ -1,254 +1,65 @@
-        #_6,                                                       \
-        _6,                                                        \
-        #_5,                                                       \
-        _5,                                                        \
-        #_4,                                                       \
-        _4,                                                        \
-        #_3,                                                       \
-        _3,                                                        \
-        #_2,                                                       \
-        _2,                                                        \
-        #_1,                                                       \
-        _1)
-#define KVLOG11(_1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11) \
-  KvLog(#_11, _11, #_10, _10, #_9, _9, #_8, _8, #_7, _7, #_6, _6, #_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG10(_1, _2, _3, _4, _5, _6, _7, _8, _9, _10) \
-  KvLog(#_10, _10, #_9, _9, #_8, _8, #_7, _7, #_6, _6, #_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG9(_1, _2, _3, _4, _5, _6, _7, _8, _9) \
-  KvLog(#_9, _9, #_8, _8, #_7, _7, #_6, _6, #_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG8(_1, _2, _3, _4, _5, _6, _7, _8) \
-  KvLog(#_8, _8, #_7, _7, #_6, _6, #_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG7(_1, _2, _3, _4, _5, _6, _7) KvLog(#_7, _7, #_6, _6, #_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG6(_1, _2, _3, _4, _5, _6) KvLog(#_6, _6, #_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG5(_1, _2, _3, _4, _5) KvLog(#_5, _5, #_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG4(_1, _2, _3, _4) KvLog(#_4, _4, #_3, _3, #_2, _2, #_1, _1)
-#define KVLOG3(_1, _2, _3) KvLog(#_3, _3, #_2, _2, #_1, _1)
-#define KVLOG2(_1, _2) KvLog(#_2, _2, #_1, _1)
-#define KVLOG1(_1) KvLog(#_1, _1)
+#define KVLOG(...) KvLog<true> KVARGS(__VA_ARGS__)
+#define ASSERT_KVLOG(...) KvLog<false> KVARGS(__VA_ARGS__)
 
 // These are recurisive variadic template functions for generating formatted key-value pairs that
 // can be attached to the end of log messages. The entrypoint is the KVLog function at the bottom
 // that only takes a variadic arg.
 //
+// The mandatory `Strict` template parameter determines whether types that do not implement `std::ostream::operator<<`
+// are allowed. If Strict=true, a static_assert will fire.
+//
 // Right now the entrypoint returns a string, because of the way our logging macros are implemented.
 // We could change them to take an extra parameter that's a stringstream and prevent the additional
 // allocation and copy from returning a string. We'll deal with that if its needed.
-template <typename K,
+template <bool Strict,
+          typename K,
           typename V,
           typename std::enable_if<concord::is_streamable<std::ostream, V>::value>::type * = nullptr>
```

Comment:
Hmm, I'm not sure I understand the use of enable_if here... Isn't there supposed to be a second template param as in enable_if<B, T> ? I don't see who sets the T in your expression :)

Reply:
It defaults to void. Since we only care that there is a type, and not what the type is to make things well formed this works.